pkg/proxy/config/config.go

```diff
@@ -369,3 +388,95 @@ func (c *ServiceConfig) handleDeleteService(obj interface{}) {
 		c.eventHandlers[i].OnServiceDelete(service)
 	}
 }
+
+// NodeMetadataConfig tracks a set of node metadata configurations.
+type NodeMetadataConfig struct {
+	listerSynced  cache.InformerSynced
+	eventHandlers []NodeMetadataHandler
+}
+
+// NewNodeMetadataConfig creates a new NodeMetadataConfig.
+func NewNodeMetadataConfig(nodeMetadataInformer informers.GenericInformer, resyncPeriod time.Duration) *NodeMetadataConfig {
+	result := &NodeMetadataConfig{
+		listerSynced: nodeMetadataInformer.Informer().HasSynced,
+	}
+
+	nodeMetadataInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    result.handleAddNodeMetadata,
+			UpdateFunc: result.handleUpdateNodeMetadata,
+			DeleteFunc: result.handleDeleteNodeMetadata,
+		},
+		resyncPeriod,
+	)
+
+	return result
+}
+
+// RegisterEventHandler registers a handler which is called on every node
+// metadata change.
+func (c *NodeMetadataConfig) RegisterEventHandler(handler NodeMetadataHandler) {
+	c.eventHandlers = append(c.eventHandlers, handler)
+}
+
+// Run waits for cache synced and invokes handlers after syncing.
+func (c *NodeMetadataConfig) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting node metadata config controller")
+
+	if !cache.WaitForNamedCacheSync("node metadata config", stopCh, c.listerSynced) {
+		return
+	}
+
+	for i := range c.eventHandlers {
+		klog.V(3).Infof("Calling handler.OnNodeMetadataSynced()")
+		c.eventHandlers[i].OnNodeMetadataSynced()
+	}
+}
+
+func (c *NodeMetadataConfig) handleAddNodeMetadata(obj interface{}) {
+	meta, ok := obj.(*metav1.PartialObjectMetadata)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+		return
+	}
+	for i := range c.eventHandlers {
+		klog.V(4).Infof("Calling handler.OnNodeMetadataAdd")
```

Comment:
same here, I think this is too verbose for v=4?

Reply:
I think add operation set to 4 is not too verbose